sd: sync to master-417-43a70e8

wbruna opened the pull request:
Another day, another giant diff 😅

We got a new file `examples/common/common.hpp`, which conflicts with `otherarch/sdcpp/common.hpp`, so I went the simplest route and placed it on a separate directory. There's also a new `version.cpp` used by `main.cpp`, but it's inert for now (it needs build support to embed the git version and commit).

And some good news: I was able to apply the LoRA using only upstream code, and discard all related Koboldcpp patches. Tested on SDXL; seems to be working so far.

Only lightly tested for now.

wbruna committed:
sd: sync to master-417-43a70e8

wbruna committed:
fix sdmain build

wbruna committed:
switch to upstream apply_loras()

wbruna commented:
BTW, this supersedes #1888 . Applying that one only makes sense if you sync llama.cpp before this PR here.

LostRuins commented:
thanks for the PR. hopefully nothing breaks too badly - will give it a test when i have a moment.

LostRuins commented:
Notice you stripped out the `std::filesystem::u8path`, we need that to handle u8 filepaths correctly cross platform like `C:\私の_models\非常好.safetensors`

Also did you check that it prevents runtime loading/switching of LoRAs from prompttags (which kcpp aims to avoid), the upstream code used to permit that e.g. `have a <lora:UltraServerCrashTime:0.9>:0.5 bad time`... and i think that part we also need to ensure that `lora_model_dir` is never set

wbruna commented:
> Notice you stripped out the `std::filesystem::u8path`, we need that to handle u8 filepaths correctly cross platform like `C:\私の_models\非常好.safetensors`

Hm... at `load_lora_model_from_file`, right? That wasn't intentional; since most of the code changed, I must have decided to restore it after the merge, and forgot. I'll restore it.

By the way, wouldn't that functionality make sense upstream, too? I can't really test it on Windows; but I could refactor everything in a way that would apply upstream, then we check it back with Koboldcpp. What do you think?

> Also did you check that it prevents runtime loading/switching of LoRAs from prompttags (which kcpp aims to avoid), the upstream code used to permit that e.g. `have a <lora:UltraServerCrashTime:0.9>:0.5 bad time`

The whole parsing code got moved out to `main.cpp` (d939f6e8 upstream); I think the `lora_model_dir` ctx parameter isn't even used anymore. So it should be safe. There is a change in behavior, though: as-is, the model will get any `<lora...>` string as part of the prompt (not sure if that's a problem).

wbruna commented:
> Hm... at `load_lora_model_from_file`, right? That wasn't intentional; since most of the code changed, I must have decided to restore it after the merge, and forgot. I'll restore it.

`StableDiffusionGGML::load_lora_model_from_file` provides the path to the `LoraModel` constructor, which calls `ModelLoader::init_from_file_and_convert_name`, which calls `ModelLoader::init_from_file`. For a `.safetensors` path, it calls `ModelLoader::init_from_safetensors_file`, which already deals with u8 paths.

I don't see the same code for gguf files, though. `stable-diffusion.cpp` _didn't_ support those for LoRAs, but it apparently does now. So maybe we should add that dance to `ModelLoader::init_from_gguf_file`? (and shouldn't we have problems with it already for model files, then?...)

LostRuins commented:
Basically whenever dealing with *any* `std::filesystem::path` on windows, it should be used. The pattern i've been copy pasting around is 
```
    #ifdef _WIN32
        std::filesystem::path fpath = std::filesystem::u8path(fname);
    #else
        std::filesystem::path fpath = std::filesystem::path(fname);
    #endif
```
its kinda duplicated in `model.cpp` and `ggml_extend.hpp`, so i'm not really sure if it's covering all possible paths yet. I've probably missed a few myself. The regular path works until you add non-english characters then it can fail.

wbruna committed:
refactor u8 path conversions and add it to the gguf reader

wbruna commented:
I moved the #ifdef dance to a helper function, and added it to the gguf reader code. And I believe all `ifstream`s now are wrapped.

But thinking about it... Couldn't we convert all those on `sdtype_adapter.cpp` instead?

LostRuins commented:
we could, assuming everywhere needed the same kind of filepath. i think some places need the stem only like the old lora method. either approach should work fine

LostRuins commented:
Alright ran through some basic tests, looks fine so i'll merge for now
